minio/credentials/providers.py

```diff
@@ -482,9 +482,8 @@ def retrieve(self):
         res = _urlopen(
             self._http_client,
             "POST",
-            self._sts_endpoint,
-            body=self._body,
-            headers={"Content-Type", "application/x-www-form-urlencoded"},
+            self._sts_endpoint + "?" + self._url, # looks like encoded url should be used to form full url, instead of passed as a body ...
```

Comment:
the encoded URI should be part of the body since its a POST request with body as per AWS S3 spec @stepanko7 do not change that.

Reply:
I get following error when passing encoded url into body:

tep@gray:~/bin$ python3 testing_minio_ldap.py 
Traceback (most recent call last):
  File "testing_minio_ldap.py", line 40, in <module>
    creds = cred_provider.retrieve()
  File "/home/step/.local/lib/python3.8/site-packages/minio-7.0.0-py3.8.egg/minio/credentials/providers.py", line 482, in retrieve
  File "/home/step/.local/lib/python3.8/site-packages/minio-7.0.0-py3.8.egg/minio/credentials/providers.py", line 72, in _urlopen
ValueError: https://localhost:9000 failed with HTTP status code 400

The testing_minio_ldap.py  script is as follows:


import os
import minio 
from minio import Minio
from minio.error import (ResponseError, BucketAlreadyOwnedByYou,
                         BucketAlreadyExists)
import urllib3
from urllib.parse import urlencode


def _urlopen(http_client, method, url, body=None, headers=None):
    """Wrapper of urlopen() handles HTTP status code."""
    res = http_client.urlopen(method, url, body=body, headers=headers)
    if res.status not in [200, 204, 206]:
        print("res: ", res.data)
        raise ValueError(
            "{0} failed with HTTP status code {1}".format(url, res.status),
        )
    return res


ca_certs = os.environ.get('SSL_CERT_FILE')
http_client = urllib3.PoolManager(
                timeout=urllib3.Timeout.DEFAULT_TIMEOUT,
                maxsize=10,
                cert_reqs='CERT_REQUIRED',
                ca_certs=ca_certs,
                retries=urllib3.Retry(
                    total=5,
                    backoff_factor=0.2,
                    status_forcelist=[500, 502, 503, 504]
                )
            )


sts_endpoint = 'https://localhost:9000'
ldap_username,ldap_password = 'user1',"User1111"

cred_provider = minio.credentials.LdapIdentityProvider(sts_endpoint,ldap_username,ldap_password,http_client)

creds = cred_provider.retrieve()

print("creds:", creds)
print("access_key:", creds.access_key)
print("secret_key:", creds.secret_key)
print("SecretToken:", creds.session_token)